### PR TITLE
Add Content Security Policy, CSRF protection and Robots policy

### PR DESCRIPTION
**Pull request by graham-willis-druid**
1036 - CSRF protection
1037 - Add CSP
1038 - Add robots (provisional file)
1039 - Add the back-button - no Javascript

### CODE REVIEW

File: packages/gafl-webapp-service/src/gafl-webapp-service.js [hunk 1]
```diff
@@ -4,5 +4,5 @@
  * Start the hapi
  */
 import { createServer, init } from './server.js'
-createServer({ port: 3000 })
+createServer({ port: 3000, debug: { request: ['error'], log: ['error'] } })
```
[the remark this answers is not in the document]
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/gafl-webapp-service.js:7]: The error message and stack trace is hidden from the console in a boom object unless you add this line. I worry that unless we do this we might lose useful error logging.

File: packages/gafl-webapp-service/src/server.js [hunk 2]
```diff
@@ -42,8 +46,42 @@ const createServer = options => {
...
+    plugin: Blankie,
+    options: {
+      /*
+       * This defines the content security policy - which is as restrictive as possible
+       * It must allow web-fonts from 'fonts.gstatic.com'
+       * Unfortunately unsafe-inline rather than script nonces must be used to prevent a console error from line
+       * 31 of the GDS template. This will probably come up as an advisory in the PEN test.
+       */
+      fontSrc: ['self', 'fonts.gstatic.com', 'data:'],
+      scriptSrc: ['self', 'unsafe-inline'],
```
[the remark this answers is not in the document]
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/server.js:66]: As I said in the comment the problem is the GDS layout template which contains an inline script. This causes a console error without in-line. It is not possible to use a nonce here without forking the template which I don't want to do. The alternative is to provide a hash of the script but the problem there is that if the template changes then the system may break and it is not obvious why.

File: packages/gafl-webapp-service/assets/robots.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+User-agent: *
+Allow: /buy
+Disallow: /buy/new
```
[the remark this answers is not in the document]
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/assets/robots.txt:3]: Yes - you are probably right. The old service does in fact disallow the entire service

[the remark this answers is not in the document]
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/assets/robots.txt:3]: I will amend that

File: packages/gafl-webapp-service/src/gafl-webapp-service.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sgardnerdell-ea) [reply packages/gafl-webapp-service/src/gafl-webapp-service.js:7]: My concern is that we do not want to log errors which are genuinely 4xx client request errors

File: packages/gafl-webapp-service/src/server.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgardnerdell-ea) [reply packages/gafl-webapp-service/src/server.js:66]: 😞 that's a shame - you'd hope that any scripts used by the templates would be referenced in a separate .js file

**Comment by graham-willis-druid** [comment]
Superceeded

**Pull request closed without merging**